cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -0,0 +1,267 @@
+      break;
+    }
+    case Type::UINT16: {
+      value_array = std::make_shared<UInt16Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT16: {
+      value_array = std::make_shared<Int16Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT32: {
+      value_array = std::make_shared<UInt32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT32: {
+      value_array = std::make_shared<Int32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::HALF_FLOAT: {
+      value_array = std::make_shared<HalfFloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::FLOAT: {
+      value_array = std::make_shared<FloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::DOUBLE: {
+      value_array = std::make_shared<DoubleArray>(tensor->size(), tensor->data());
+      break;
+    }
+    default: {
+      return Status::NotImplemented("Unsupported tensor type: ",
+                                    tensor->type()->ToString());
+    }
+  }
+  arr = std::make_shared<FixedSizeListArray>(ext_type->storage_type(), tensor->shape()[0],
+                                             value_array);
+  auto ext_data = arr->data();
+  ext_data->type = ext_type;
+  return MakeArray(ext_data);
+}
+
+Result<std::shared_ptr<Tensor>> FixedShapeTensorType::ToTensor(
+    std::shared_ptr<Array> arr) const {
+  // To convert an array of n dimensional tensors to a n+1 dimensional tensor we
+  // interpret the array's length as the first dimension the new tensor. Further, we
+  // define n+1 dimensional tensor's strides by front appending a new stride to the n
+  // dimensional tensor's strides.
+
+  ARROW_DCHECK_EQ(arr->null_count(), 0) << "Null values not supported in tensors.";
+  auto ext_arr = internal::checked_pointer_cast<FixedSizeListArray>(
+      internal::checked_pointer_cast<ExtensionArray>(arr)->storage());
+
+  std::vector<int64_t> shape = shape_;
+  shape.insert(shape.begin(), 1, arr->length());
+
+  std::vector<int64_t> tensor_strides = strides();
+  tensor_strides.insert(tensor_strides.begin(), 1, arr->length() * tensor_strides[0]);
+
+  std::shared_ptr<Buffer> buffer = ext_arr->values()->data()->buffers[1];
+  return *Tensor::Make(ext_arr->value_type(), buffer, shape, tensor_strides, dim_names());
+}
+
+std::shared_ptr<FixedShapeTensorType> fixed_shape_tensor(
+    const std::shared_ptr<DataType>& value_type, const std::vector<int64_t>& shape,
+    const std::vector<int64_t>& permutation, const std::vector<std::string>& dim_names) {
+  ARROW_CHECK(is_tensor_supported(value_type->id()));
```

Comment:
Is it needed to restrict creating this type to just numeric types? (that's certainly the most typical usecase, but it's not necessarily needed to strictly require that? And if we want to require that, that should be part of the spec, I think?)